client/allocrunner/taskrunner/plugin_supervisor_hook.go

```diff
@@ -230,19 +233,23 @@ func (h *csiPluginSupervisorHook) setSocketHook() {
 	h.socketPath = filepath.Join(h.socketMountPoint, structs.CSISocketName)
 }
 
-// Poststart is called after the task has started. Poststart is not
-// called if the allocation is terminal.
+// Poststart is called after the task has started (or restarted). Poststart is
+// not called if the allocation is terminal.
 //
 // The context is cancelled if the task is killed.
-func (h *csiPluginSupervisorHook) Poststart(_ context.Context, _ *interfaces.TaskPoststartRequest, _ *interfaces.TaskPoststartResponse) error {
-
-	// If we're already running the supervisor routine, then we don't need to try
-	// and restart it here as it only terminates on `Stop` hooks.
-	h.runOnce.Do(func() {
-		h.setSocketHook()
-		go h.ensureSupervisorLoop(h.shutdownCtx)
-	})
+func (h *csiPluginSupervisorHook) Poststart(ctx context.Context, _ *interfaces.TaskPoststartRequest, _ *interfaces.TaskPoststartResponse) error {
+
+	// If we're already running the supervisor routine, then we don't need to
+	// try and restart it here as it only terminates on `Stop` hooks and health
+	// timeouts (which restart the task)
+	h.supervisorIsRunningLock.Lock()
+	h.supervisorIsRunningLock.Unlock()
```

Comment:
was this meant to be

```suggestion
	h.supervisorIsRunningLock.Lock()
	defer h.supervisorIsRunningLock.Unlock()
```

?

Reply:
:blush: that's embarassing